src/rwlock/futures.rs

```diff
@@ -0,0 +1,300 @@
+    type Output = RwLockWriteGuard<'a, T>;
+
+    #[inline]
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        ready!(Pin::new(&mut self.raw).poll(cx));
+
+        Poll::Ready(RwLockWriteGuard {
+            lock: self.raw.lock,
+            value: self.value,
+        })
+    }
+}
+
+/// The future returned by [`RwLock::write_arc`].
+pub struct WriteArc<'a, T: ?Sized> {
+    /// Raw write lock acquisition future, doesn't depend on `T`.
+    pub(super) raw: RawWrite<'a>,
+
+    pub(super) lock: &'a Arc<RwLock<T>>,
+}
+
+unsafe impl<T: Send + Sync + ?Sized> Send for WriteArc<'_, T> {}
+unsafe impl<T: Send + Sync + ?Sized> Sync for WriteArc<'_, T> {}
+
+impl<T: ?Sized> fmt::Debug for WriteArc<'_, T> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("WriteArc { .. }")
+    }
+}
+
+impl<T: ?Sized> Unpin for WriteArc<'_, T> {}
+
+impl<'a, T: ?Sized> Future for WriteArc<'a, T> {
+    type Output = ArcRwLockWriteGuard<T>;
+
+    #[inline]
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        ready!(Pin::new(&mut self.raw).poll(cx));
+
+        Poll::Ready(ArcRwLockWriteGuard {
+            lock: self.lock.clone(),
+        })
+    }
+}
+
+/// The future returned by [`RwLockUpgradableReadGuard::upgrade`].
+pub struct Upgrade<'a, T: ?Sized> {
+    /// Raw read lock upgrade future, doesn't depend on `T`.
+    pub(super) raw: RawUpgrade<'a>,
+
+    /// Pointer to the value protected by the lock. Invariant in `T`.
+    pub(super) value: *mut T,
+}
+
+impl<T: ?Sized> fmt::Debug for Upgrade<'_, T> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Upgrade").finish()
+    }
+}
+
+impl<T: ?Sized> Unpin for Upgrade<'_, T> {}
+
+impl<'a, T: ?Sized> Future for Upgrade<'a, T> {
+    type Output = RwLockWriteGuard<'a, T>;
+
+    #[inline]
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let lock = ready!(Pin::new(&mut self.raw).poll(cx));
+
+        Poll::Ready(RwLockWriteGuard {
+            lock,
+            value: self.value,
+        })
+    }
+}
+
+/// The future returned by [`ArcRwLockUpgradableReadGuard::upgrade`].
+pub struct ArcUpgrade<T: ?Sized> {
+    /// Raw read lock upgrade future, doesn't depend on `T`.
+    /// `'static` is a lie, this field is actually referencing the
+    /// `Arc` data. But since this struct also stores said `Arc`, we know
+    /// this value will be alive as long as the struct is.
+    ///
+    /// Yes, one field of the `ArcUpgrade` struct is referencing another.
+    /// Such self-references are usually not sound without pinning.
+    /// However, in this case, there is an indirection via the heap;
+    /// moving the `ArcUpgrade` won't move the heap allocation of the `Arc`,
+    /// so the reference inside `RawUpgrade` isn't invalidated.
+    pub(super) raw: RawUpgrade<'static>,
```

Comment:
Overall this seems sound, but this field in particular should be wrapped in a `ManuallyDrop` and explicitly dropped before the `lock` field is. That way there's no chance for a dangling reference.